File: backend/train/src/main/java/com/example/train/services/TicketService.java
package com.example.train.services;

import java.util.List;

import com.example.train.dto.request.TicketRequestDto;
import com.example.train.dto.response.PageResponse;
import com.example.train.dto.response.TicketDetailResponse;
import com.example.train.entity.Ticket;
import com.example.train.entity.TicketStatus;

public interface TicketService {
    List<Ticket> addTickets(List<TicketRequestDto> ticketRequestDtos);
    void updateTicket(int id, TicketRequestDto TicketRequestDto);
    void deleteTicket(int id);
    void restoreTicket(int id);

    void updateTicketStatus(int id, TicketStatus status);

    PageResponse<?> getAllTicketAndSearchWithPagingAndSorting(int pageNo, int pageSize, String search, String sortBy,int id);

   // List<TicketDetailResponse> getTicketsByUserId(int userId, TicketStatus status);
}

File: frontend/src/app/_services/ticket.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { inject, Injectable } from '@angular/core';
import { environment } from '../../environments/environment.development';
import { Ticket, TicketStatus } from '../_models/ticket.module';
import { Observable } from 'rxjs';

@Injectable({
  providedIn: 'root',
})
export class TicketService {
  private http = inject(HttpClient);
  private baseUrl = environment.apiUrl;
  private ticketData: any;
  private personalData: any;

  setPersonalData(data: any): void{
    this.personalData = data;
    localStorage.setItem('personalData', JSON.stringify(this.personalData));
  }
  getPersonalData(): any{
    return localStorage.getItem('personalData');
  }
  setTicketData(data: any): void {
    this.ticketData = data;
    localStorage.setItem('ticketData', JSON.stringify(this.ticketData));

  }

  getTicketData(): any {
    return localStorage.getItem('ticketData');
  }

  getWithLimit(
    page: number = 1,
    size: number = 10,
    search: string = '',
    sortBy: string = 'id,desc',
    id:number
  ) {
    let params = new HttpParams();
    params = params.set('pageNumber', page.toString());
    params = params.set('pageSize', size.toString());
    params = params.set('sortBy', sortBy);

    if (search) {
      params = params.set('search', search);
    }

    return this.http.get<any>(`${this.baseUrl}/ticket/list/${id}`, {
      observe: 'response',
      params,
    });
  }
  getTicketById(id: number) {
    return this.http.get<{ status: number; message: string; data: Ticket }>(`${this.baseUrl}/ticket/${id}`);
  }

  addTickets(tickets: Ticket[]): Observable<any> {
    return this.http.post(`${this.baseUrl}/ticket/add`, tickets);
  }

  deleteTicket(id: number): Observable<any> {
    return this.http.delete(`${this.baseUrl}/ticket/delete/${id}`, {
      headers: { 'Content-Type': 'application/json' }
    });
  }
  updateTicketStatus(id: number, status: TicketStatus): Observable<any> {
    console.log(status);
    return this.http.put(`${this.baseUrl}/ticket/updateStatus/${id}/${status}`, {}  );
  }

  getTicketsByUserId(id: number, status: TicketStatus) {
    return this.http.get<any>(`${this.baseUrl}/ticket/list/${id}/${status}`);
  }

  getPaidSeats(){
    return this.http.get<any>(`${this.baseUrl}/ticket/paidSeats`);
  }
}
